services/controlplane/features/client.go

```diff
@@ -0,0 +1,147 @@
+package features
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"runtime"
+	"sort"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+)
+
+var (
+	DefaultTimeout = 30 * time.Second
+	MaxRetries     = uint64(3)
+)
+
+type OptFn func(c *Client)
+
+func WithURL(url string) OptFn {
+	return func(c *Client) {
+		c.url = url
+	}
+}
+
+func WithHTTPClient(httpClient *http.Client) OptFn {
+	return func(c *Client) {
+		c.client = httpClient
+	}
+}
+
+func WithTimeout(timeout time.Duration) OptFn {
+	return func(c *Client) {
+		c.client.Timeout = timeout
+	}
+}
+
+type Client struct {
+	client   *http.Client
+	ua       string
+	url      string
+	identity identity.Identifier
+}
+
+type payload struct {
+	Components []component `json:"components"`
+}
+
+type component struct {
+	Name     string   `json:"name"`
+	Features []string `json:"features"`
+}
+
+func hostname() string {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "unknown"
+	}
+	return hostname
+}
+
+func New(identity identity.Identifier, fns ...OptFn) *Client {
+	c := &Client{
+		client: &http.Client{
+			Timeout: DefaultTimeout,
+		},
+		url:      config.GetEnv("CONFIG_BACKEND_URL", "https://api.rudderlabs.com"),
```

Comment:
Can we please avoid spreading the config package everywhere and have dependencies properly injected? We're duplicating also the default (i.e. `https://api.rudderlabs.com`) each time. It would be great if we could have `url` as a parameter since it's not optional and remove the `WithURL` and return an error with the provided `url` is empty.